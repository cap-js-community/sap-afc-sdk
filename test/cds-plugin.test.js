"use strict";

const cds = require("@sap/cds");

const { authorization, clearEventQueue } = require("./helper");
const { config: eventQueueConfig } = require("@cap-js-community/event-queue");

const log = cds.test.log();

const { GET, POST, axios, test } = cds.test(__dirname + "/..");

process.env.PORT = 0; // Random

cds.env.requires["sap-afc-sdk"].broker = true;
cds.env.requires["sap-afc-sdk"].ui.path = "/custom";
cds.env.requires["sap-afc-sdk"].broker = {
  SBF_SERVICE_CONFIG: {
    service: {
      extend_credentials: {
        shared: {
          endpoints: {
            api: "/api",
            "job-scheduling-v1": "/api/job-scheduling/v1",
          },
          "oauth2-configuration": {
            "credential-types": ["binding-secret", "x509"],
          },
        },
      },
      extend_xssecurity: {
        shared: {
          "oauth2-configuration": {
            "credential-types": ["binding-secret", "x509"],
          },
        },
      },
    },
  },
  SBF_BROKER_CREDENTIALS_HASH: {
    "broker-user": "sha256:E/VwOIZC2ZKPY+VTPCg1Mfu/bFDlxgkk9AgyrzhgpHs=:WNlnGu7jfxHGwy14BoVEiFXzp9RcwGiInauL4iOTe5E=",
  },
};

describe("CDS Plugin", () => {
  beforeEach(async () => {
    axios.defaults.headers = {
      Authorization: authorization.zeus,
    };
    await clearEventQueue();
    await test.data.reset();
  });

  it("Boostrap", async () => {
    expect(log.output).toEqual(
      expect.stringMatching(
        /\[\/broker] - Failed to start broker AssertionError \[ERR_ASSERTION]: Could not find credentials service/s,
      ),
    );
  });

  it("GET Welcome Page", async () => {
    const response = await GET("/");
    response.data = response.data.replace(/>@cap-js-community\/sap-afc-sdk (.*?)</, ">@cap-js-community/sap-afc-sdk<");
    expect(response.status).toBe(200);
    expect(response.data).toMatchSnapshot();
  });

  it("GET API Root", async () => {
    const response = await GET("/api");
    expect(response.status).toBe(200);
    for (const name in response.data) {
      response.data[name] = response.data[name].replace(/localhost:\d+/, "localhost:4004");
    }
    expect(response.data).toMatchSnapshot();
  });

  it("GET App Config", async () => {
    const response = await GET("/appconfig/fioriSandboxConfig.json");
    expect(response.status).toBe(200);
    expect(response.data).toMatchSnapshot();
  });

  it("GET Feature Toggles", async () => {
    const response = await GET("/rest/feature/state()");
    expect(response.status).toBe(200);
    expect(response.data).toMatchSnapshot();
  });

  it("POST Feature Toggles", async () => {
    let response = await POST("/rest/feature/redisUpdate", {
      key: "eventQueue/active",
      value: false,
    });
    expect(response.status).toBe(204);
    eventQueueConfig.isEventQueueActive = true;
    response = await POST("/rest/feature/redisUpdate", {
      key: "eventQueue/instanceLoadLimit",
      value: 11,
    });
    expect(response.status).toBe(204);
    eventQueueConfig.instanceLoadLimit = 11;
    response = await GET("/rest/feature/state()");
    expect(response.data).toMatchSnapshot();
  });

  it("POST Feature Toggles (remove)", async () => {
    let response = await POST("/rest/feature/redisUpdate", {
      key: "eventQueue/active",
      value: null,
      options: {
        clearSubScopes: true,
      },
    });
    expect(response.status).toBe(204);
    eventQueueConfig.isEventQueueActive = false;
    response = await POST("/rest/feature/redisUpdate", {
      key: "eventQueue/instanceLoadLimit",
      value: null,
      options: {
        clearSubScopes: true,
      },
    });
    expect(response.status).toBe(204);
    eventQueueConfig.instanceLoadLimit = 5;
    response = await GET("/rest/feature/state()");
    expect(response.data).toMatchSnapshot();
  });
});
