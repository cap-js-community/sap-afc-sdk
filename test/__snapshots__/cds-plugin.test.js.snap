// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDS Plugin GET API Root 1`] = `
{
  "/api/job-scheduling/v1": "http://localhost:4004/api/job-scheduling/v1",
}
`;

exports[`CDS Plugin GET App Config 1`] = `
{
  "services": {
    "ClientSideTargetResolution": {
      "adapter": {
        "config": {
          "inbounds": {
            "JobMonitor": {
              "action": "monitor",
              "resolutionResult": {
                "additionalInformation": "SAPUI5.Component=scheduling.monitoring.job",
                "applicationType": "SAPUI5",
                "url": "/custom/./scheduling.monitoring.job/webapp",
              },
              "semanticObject": "Job",
              "signature": {
                "additionalParameters": "allowed",
                "parameters": {
                  "Job.ID": {
                    "renameTo": "ID",
                  },
                  "Job.parameters.ID": {
                    "renameTo": "ID",
                  },
                },
              },
              "title": "Monitor Scheduling Jobs",
            },
          },
        },
      },
    },
    "LaunchPage": {
      "adapter": {
        "config": {
          "catalogs": [],
          "groups": [
            {
              "id": "Scheduling",
              "isGroupLocked": false,
              "isPreset": true,
              "isVisible": true,
              "tiles": [
                {
                  "id": "JobMonitor",
                  "properties": {
                    "targetURL": "#Job-monitor",
                    "title": "Monitor Scheduling Jobs",
                  },
                  "tileType": "sap.ushell.ui.tile.StaticTile",
                },
              ],
              "title": "Scheduling",
            },
          ],
        },
      },
    },
    "NavTargetResolution": {
      "config": {
        "enableClientSideTargetResolution": true,
      },
    },
  },
}
`;

exports[`CDS Plugin GET Feature Toggles 1`] = `
{
  "eventQueue/active": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": true,
  },
  "eventQueue/instanceLoadLimit": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 100,
  },
}
`;

exports[`CDS Plugin GET Welcome Page 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <title>@cap-js-community/sap-afc-sdk</title>
        <meta name="color-scheme" content="dark light">
        <style>
body {
  font-family: Avenir Next, sans-serif;
  background: #f8fbff;
}

body #welcome {
  margin: 30px auto 50px;
  padding: 0 30px;
  max-width: 900px;
}

body #welcome h1, #welcome h3, #welcome h2, #welcome p {
  margin: 20px 0;
  color: #47637d;
  font-weight: 400;
}

body #welcome h3.header {
  margin: 20px 0 0 0;
  background-color: #537492;
  color: white;
}

body #welcome h3 a {
  color: inherit;
  font-weight: 500;
}

body #welcome h3 a span {
  padding: 5px 15px;
  font-weight: 300;
  transition: all 0.2s ease-in;
}

body #welcome h3 a:first-child span {
  display: inline-block;
  padding: 5px 15px;
}

body #welcome h3 a span:hover {
  color: #92b2d0;
}

body #welcome ul {
  margin: 0;
  background: #91adc5;
  list-style-type: none;
  padding-inline-start: 20px;
}

body #welcome li {
  display: inline-grid;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  width: 100%;
  margin-bottom: 1px;
  background-color: #d0e9ff;
  padding: 0;
  /*
  -moz-transition: background color 0.2s ease-in;
  -webkit-transition: background 0.2s ease-in;
  -o-transition: background 0.2s ease-in;
  /* Standard */
  transition: all 0.1s ease-in;
}

body #welcome ul li div:not(:first-child) a {
  font-weight: 300;
}

body #welcome a {
  text-decoration: none;
}

body #welcome li a, body #welcome li > span {
  padding: 4px 10px;
  font-weight: 500;
  color: #1d5985;
  transition: all 0.1s ease-in;
}
body #welcome li.operation span {
  font-style: italic;
}

body #welcome li a span:hover {
  color: #f0faff;
  background-color: #a7c8e1;
}

body #welcome .preview {
  font-size: small;
}

footer {
  border-top: .5px solid;
  margin-top: 44px;
  padding-top: 22px;
  width: 400px;
  font-size: 90%;
}

@media (prefers-color-scheme: dark) {
  body { background: #0a2138; }

  body #welcome h1, #welcome h3, #welcome h2, #welcome p {
      color: #a6c1d9;
  }
  body #welcome h3 a {
      color: #e2efff;
  }
  body #welcome ul {
      background: #7894ad;
  }
  body #welcome li {
      background-color: #506d88;
  }
  body #welcome li a, body #welcome li:hover div:not(:first-child) a, body #welcome li > span {
      color: #d0e4f2;
  }
  body #welcome li a:hover {
      color: white;
      background-color: #7894ad;
  }

  body #welcome h3.header {
      background-color: #304c65;
  }

  footer {
      border-top: .5px solid #456;
      color: #567;
  }
}
        </style>
    </head>
    <body>
        <div id="welcome">
            <h1> Welcome to <i>@sap/cds</i> Server </h1>
            <p class="subtitle"> Serving @cap-js-community/sap-afc-sdk</p>

            <p> These are the paths currently served:</p>

            <h2> Web Applications: </h2>
            <ul>
                <li><a href="launchpad.html"><span>/launchpad.html</span></a></li>
<li><a href="scheduling.monitoring.job/webapp/index.html"><span>/scheduling.monitoring.job/webapp</span></a></li>
            </ul>

            <h2> Service Endpoints: </h2>
            
      <div id="featureservice-rest">
        <h3 class="header">
          <a href="/rest/feature"><span>/rest/feature</span></a> 
        </h3>
        <ul>
        </ul>
        <ul>
          <li id="featureservice-rest-state" class="operation">
            <div>
              <a href="/rest/feature/state" title="/rest/feature/state"><span>state ()</span></a>
            </div>
          </li>
        </ul>
      </div>
  
      <div id="schedulingwebsocketservice-ws">
        <h3 class="header">
          <a href="/ws/job-scheduling"><span>/ws/job-scheduling</span></a> 
        </h3>
        <ul>
        </ul>
        <ul>
        </ul>
      </div>
  
      <div id="schedulingmonitoringservice-odata">
        <h3 class="header">
          <a href="/odata/v4/job-scheduling/monitoring"><span>/odata/v4/job-scheduling/monitoring</span></a><span>/</span> <a href="/odata/v4/job-scheduling/monitoring/$metadata"><span>$metadata</span></a> 
        </h3>
        <ul>
          <li id="schedulingmonitoringservice-odata-jobdefinition">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobDefinition"><span>JobDefinition</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobDefinition#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-jobparameterdefinition">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobParameterDefinition"><span>JobParameterDefinition</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobParameterDefinition#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-job">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/Job"><span>Job</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/Job#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-jobparameter">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobParameter"><span>JobParameter</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobParameter#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-jobresult">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobResult"><span>JobResult</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobResult#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-jobresultmessage">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobResultMessage"><span>JobResultMessage</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobResultMessage#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-jobstatus">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/JobStatus"><span>JobStatus</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/JobStatus#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-resulttype">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/ResultType"><span>ResultType</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/ResultType#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-parametertype">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/ParameterType"><span>ParameterType</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/ParameterType#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-datatype">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/DataType"><span>DataType</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/DataType#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-mappingtype">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/MappingType"><span>MappingType</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/MappingType#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
          <li id="schedulingmonitoringservice-odata-messageseverity">
            <div>
              <a href="/odata/v4/job-scheduling/monitoring/MessageSeverity"><span>MessageSeverity</span></a>
            </div>
            <div><a href="/$fiori-preview/SchedulingMonitoringService/MessageSeverity#preview-app"><span class="preview" title="Preview in Fiori elements">Fiori preview</span></a></div>
          </li>
        </ul>
        <ul>
        </ul>
      </div>
  
      <div id="schedulingproviderservice-rest">
        <h3 class="header">
          <a href="/api/job-scheduling/v1"><span>/api/job-scheduling/v1</span></a> <span><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></span>
        </h3>
        <ul>
          <li id="schedulingproviderservice-rest-jobdefinition">
            <div>
              <a href="/api/job-scheduling/v1/JobDefinition"><span>JobDefinition</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
          <li id="schedulingproviderservice-rest-jobparameterdefinition">
            <div>
              <a href="/api/job-scheduling/v1/JobParameterDefinition"><span>JobParameterDefinition</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
          <li id="schedulingproviderservice-rest-job">
            <div>
              <a href="/api/job-scheduling/v1/Job"><span>Job</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
          <li id="schedulingproviderservice-rest-jobparameter">
            <div>
              <a href="/api/job-scheduling/v1/JobParameter"><span>JobParameter</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
          <li id="schedulingproviderservice-rest-jobresult">
            <div>
              <a href="/api/job-scheduling/v1/JobResult"><span>JobResult</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
          <li id="schedulingproviderservice-rest-jobresultmessage">
            <div>
              <a href="/api/job-scheduling/v1/JobResultMessage"><span>JobResultMessage</span></a>
            </div>
            <div><a href="/api-docs/api/job-scheduling/v1"><span class="preview" title="Show in Swagger UI">Open API</span></a></div>
          </li>
        </ul>
        <ul>
        </ul>
      </div>
  

            <footer>
                This is an automatically generated page. <br>
                You can replace it with a custom <code>./app/index.html</code>.
            </footer>
        </div>

    </body>
</html>
"
`;

exports[`CDS Plugin POST Feature Toggles (remove) 1`] = `
{
  "eventQueue/active": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": true,
  },
  "eventQueue/instanceLoadLimit": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 100,
  },
}
`;

exports[`CDS Plugin POST Feature Toggles 1`] = `
{
  "eventQueue/active": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": true,
    "rootValue": false,
  },
  "eventQueue/instanceLoadLimit": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 100,
    "rootValue": 11,
  },
}
`;
